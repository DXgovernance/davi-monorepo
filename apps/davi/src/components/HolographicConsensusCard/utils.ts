import { BigNumber, FixedNumber } from 'ethers';
import { Proposal } from 'types/types.guilds.d';
import { IStakes } from './types';

export const calculateSpeedometerValue = (
  totalStaked: Proposal['totalStaked']
): number => {
  if (!totalStaked) return 0;

  const stakedAgainst = totalStaked[0];
  const stakedFor = totalStaked[1];

  const stakeDifference = stakedFor.sub(stakedAgainst);
  let percentage: number;

  if (stakeDifference.gt(0)) {
    percentage = stakeDifference.mul(100).div(stakedFor).toNumber();
  } else if (stakeDifference.eq(0)) {
    percentage = 0;
  } else {
    percentage = stakeDifference.mul(100).div(stakedAgainst).toNumber();
  }
  return 5000 + percentage * 50;
};

export const calculateBNPercentage = (
  value: BigNumber,
  percentage: string | number
) => {
  // percentage should be an integer (before being divided by 100)
  // if a percentage has decimals, it trims them
  // TODO: handle decimals?
  if (typeof percentage === 'string') {
    if (percentage.includes('.')) {
      const percentageSplit = percentage.split('.');
      percentage = percentageSplit[0];
    }
  }
  if (typeof percentage === 'number') {
    if (percentage % 1 !== 0) {
      percentage = Math.trunc(percentage);
    }
  }

  const valueFixed = FixedNumber.fromValue(value);
  const percentageFixed = FixedNumber.fromValue(BigNumber.from(percentage));
  const hundredFixed = FixedNumber.fromValue(BigNumber.from(100));

  const resultFixed = valueFixed
    .mulUnsafe(percentageFixed)
    .divUnsafe(hundredFixed);

  const resultFixedInteger = resultFixed.toString().split('.')[0];
  const resultBN = BigNumber.from(resultFixedInteger);
  return resultBN;
};

export const checkUserStakeOption = (
  userAddress: string,
  stakeDetails: IStakes
) => {
  const userStakeFor = stakeDetails.for.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeFor) return 'for';

  const userStakeAgainst = stakeDetails.against.find(
    stake => stake.staker.toLowerCase() === userAddress.toLowerCase()
  );
  if (userStakeAgainst) return 'against';

  return null;
};

export const calculatePotentialReward = (
  stakes: IStakes,
  userCurrentStake: BigNumber,
  userAddress: string,
  userOption: BigNumber,
  totalStaked: Proposal['totalStaked'],
  daoBounty: BigNumber
) => {
  if (userOption === undefined || userOption === null) return BigNumber.from(0);
  const userStakedOptionIndex: number = userOption.toNumber() - 1;
  let userStakedOptionString: string;
  if (userStakedOptionIndex === 0) userStakedOptionString = 'against';
  if (userStakedOptionIndex === 1) userStakedOptionString = 'for';
  if (!userStakedOptionString)
    throw new Error('Invalid userStakedOptionString');

  const stakedInUserVote: BigNumber =
    totalStaked[userStakedOptionIndex].add(userCurrentStake);
  const stakedInUserVoteFixed: FixedNumber =
    FixedNumber.fromValue(stakedInUserVote);

  if (stakedInUserVote.isZero()) return BigNumber.from(0);

  const totalStakedWithoutDaoBounty = userCurrentStake
    .add(totalStaked[0])
    .add(totalStaked[1])
    .sub(daoBounty);

  // Calculate what the user staked
  const previouslyStakedByUser = stakes[userStakedOptionString].reduce(
    (acc: BigNumber, cur) => {
      if (cur.staker.toLowerCase() === userAddress.toLowerCase()) {
        return acc.add(cur.amount);
      }
      return acc;
    },
    BigNumber.from(0)
  );

  const totalUserStake = previouslyStakedByUser.add(userCurrentStake);

  // Calculate the coefficient for the base reward

  const initialRewardCoefficient = FixedNumber.fromValue(
    totalStakedWithoutDaoBounty
  ).divUnsafe(stakedInUserVoteFixed);

  const initialReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    initialRewardCoefficient
  );

  const initialRewardBN = BigNumber.from(
    initialReward.toString().split('.')[0]
  );

  // If the user voted for "NO", then the reward is just the initial reward
  if (userOption.eq(1)) return initialRewardBN;

  // If the user voted for "YES", then the reward is the initial reward + a percentage of the DAO bounty
  const daoBountyRewardCoefficient = FixedNumber.fromValue(daoBounty).divUnsafe(
    stakedInUserVoteFixed
  );

  const daoBountyReward = FixedNumber.fromValue(totalUserStake).mulUnsafe(
    daoBountyRewardCoefficient
  );

  const daoBountyRewardBN = BigNumber.from(
    daoBountyReward.toString().split('.')[0]
  );

  return initialRewardBN.add(daoBountyRewardBN);
};
