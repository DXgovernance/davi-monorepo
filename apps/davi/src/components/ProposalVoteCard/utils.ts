import { VoteOnProposalProps } from './types';

export const checkVotingPower = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};
