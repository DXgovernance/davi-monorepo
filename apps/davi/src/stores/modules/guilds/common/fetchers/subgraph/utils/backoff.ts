import { useBlockNumber } from 'wagmi';

export const useBackoff = (refetch: () => any) => {
  const { data: block } = useBlockNumber({
    watch: true,
  });
  let currentBlock = 0;

  // Backoff and retry logic
  const backoff = (fun, successFun, failureFun, exponent) => {
    setTimeout(async () => {
      const { data } = await fun();
      if (currentBlock < 1) {
        console.log('Still fetching current block');
      } else if (data?._meta.block.number > currentBlock) {
        successFun();
      } else if (exponent <= 20) {
        backoff(fun, successFun, failureFun, exponent + 1);
      } else {
        failureFun();
      }
    }, Math.pow(2, exponent) + Math.random() * 10000);
  };

  return async () => {
    currentBlock = block;
    backoff(
      refetch,
      () => (currentBlock = 0),
      () => console.log('Failed to fetch new data'),
      0
    );
  };
};
