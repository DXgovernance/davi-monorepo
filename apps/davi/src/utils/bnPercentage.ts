import { BigNumber } from 'ethers';

const truncateNumber = (number: number, decimals: number): number => {
  if (Number.isInteger(number)) return number;

  const stringResult = number.toString();
  const dotPosition = stringResult.indexOf('.');

  return Number(stringResult.slice(0, dotPosition + decimals + 1));
};

const getAmountOfDecimalPlaces = (number: number): number => {
  if (Number.isInteger(number)) return 0;
  const numberString = number.toString();
  const dotPosition = numberString.indexOf('.');
  const numberOfDecimals = numberString.length - dotPosition - 1;
  return numberOfDecimals;
};

// Used to calculate percentages of big numbers due to decimal limitation for things like voting power
export function getBigNumberPercentage(
  amount: BigNumber,
  totalAmount: BigNumber,
  decimals: number = 2
) {
  if (!amount || !totalAmount) return null;
  if (totalAmount.isZero()) return 0;

  const percent = amount
    .mul(100)
    .mul(Math.pow(10, decimals + 1))
    .div(totalAmount);

  let result = Math.round(percent.toNumber()) / Math.pow(10, decimals + 1);

  const numberOfDecimals = getAmountOfDecimalPlaces(result);
  if (numberOfDecimals < decimals + 1) return result;

  const lastDigit = parseInt(result.toString().slice(-1));
  if (lastDigit >= 5) result += 1 / Math.pow(10, decimals);

  return truncateNumber(result, decimals);
}
