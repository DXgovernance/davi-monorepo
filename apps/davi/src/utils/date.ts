import moment from 'moment';
import { BigNumber } from './bignumber';

export const formatDate = (date: Date) => {
  return moment(date).format('DD.MM - HH:mm');
};

export const timestampToDate = (timestamp: BigNumber) => {
  return moment.unix(timestamp.toNumber()).format('YYYY.DD.MM - HH:mm');
};

export const timeToTimestamp = (timestamp: BigNumber) => {
  if (timestamp.toNumber() > moment().unix())
    return moment().to(moment.unix(timestamp.toNumber())).toString();
  else return '';
};

export const addZero = value => {
  return value > 9 ? value : `0${value}`;
};
