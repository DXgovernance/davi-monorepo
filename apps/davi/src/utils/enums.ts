export enum Web3Errors {
  UNKNOWN_ERROR,
  SIGNATURE_REJECTED,
}
export enum TXEvents {
  TX_HASH = 'txhash',
  RECEIPT = 'receipt',
  CONFIRMATION = 'confirmation',
  TX_ERROR = 'txerror',
  FINALLY = 'finally',
  INVARIANT = 'invariant',
}

export enum VoteDecision {
  None,
  YES,
  NO,
}

export enum WalletSchemeProposalState {
  None,
  Submitted,
  Rejected,
  ExecutionSucceded,
  ExecutionTimeout,
}

export enum VotingMachineProposalState {
  None,
  ExpiredInQueue,
  Executed,
  Queued,
  PreBoosted,
  Boosted,
  QuietEndingPeriod,
  Rejected,
  Passed,
}

export enum PendingAction {
  'None',
  'Boost',
  'Execute',
  'Finish',
  'Redeem',
  'RedeemForBeneficiary',
}

export const enumKeys = <O extends object, K extends keyof O = keyof O>(
  obj: O
): K[] => Object.keys(obj).filter(k => Number.isNaN(+k)) as K[];
