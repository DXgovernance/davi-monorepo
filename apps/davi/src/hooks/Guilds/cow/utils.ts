import { GNOSIS_ID, LOCALHOST_ID, MAINNET_ID } from 'utils';

export const isAvailableOnCow = (chainId: number) => {
  return [MAINNET_ID, GNOSIS_ID, LOCALHOST_ID].includes(chainId);
};
