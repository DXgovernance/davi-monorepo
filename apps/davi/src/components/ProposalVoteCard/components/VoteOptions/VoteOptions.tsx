import { Bullet } from 'components/primitives/Bullet';
import { Loading } from 'components/primitives/Loading';
import { useTheme } from 'styled-components';
import { useTranslation } from 'react-i18next';

import { getOptionLabel } from 'components/ProposalVoteCard/utils';
import {
  ResultRowProps,
  VoteResultsProps,
} from 'components/ProposalVoteCard/types';
import {
  VotesRowWrapper,
  VoteOption,
  OptionBullet,
} from './VoteOptions.styled';

const VoteOptionsRow: React.FC<ResultRowProps> = ({
  optionKey,
  proposalMetadata,
}) => {
  const theme = useTheme();
  const { t } = useTranslation();

  const isReady = optionKey !== undefined;

  const label = getOptionLabel({ metadata: proposalMetadata, optionKey, t });
  return (
    <VotesRowWrapper>
      <VoteOption>
        <OptionBullet>
          {isReady ? (
            <Bullet color={theme?.colors?.votes?.[optionKey]} size={8} />
          ) : (
            <Loading
              loading
              text
              skeletonProps={{ circle: true, height: 16, width: 16 }}
            />
          )}
        </OptionBullet>
        {isReady ? label : <Loading loading text />}
      </VoteOption>
    </VotesRowWrapper>
  );
};

const VoteOptions: React.FC<VoteResultsProps> = ({
  // isPercent,
  options,
  proposalMetadata,
}) => {
  const orderedOptions = options && [...Object.keys(options).slice(1), '0'];

  return orderedOptions ? (
    <>
      {orderedOptions.map(key => (
        <VoteOptionsRow
          key={key}
          optionKey={Number(key)}
          // isPercent={isPercent}
          options={options}
          proposalMetadata={proposalMetadata}
        />
      ))}
    </>
  ) : (
    <>
      <VoteOptionsRow
        // isPercent={isPercent}
        options={options}
        proposalMetadata={proposalMetadata}
      />
      <VoteOptionsRow
        // isPercent={isPercent}
        options={options}
        proposalMetadata={proposalMetadata}
      />
    </>
  );
};

export default VoteOptions;
