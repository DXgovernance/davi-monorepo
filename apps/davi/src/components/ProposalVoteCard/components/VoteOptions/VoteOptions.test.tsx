import { render } from 'utils/tests';
import VoteOptions from './VoteOptions';
import { mockVoteResults } from '../../fixture';

describe('VoteResults', () => {
  it('matches the snapshot', () => {
    const { container } = render(<VoteOptions {...mockVoteResults} />);
    expect(container).toMatchSnapshot();
  });
});
