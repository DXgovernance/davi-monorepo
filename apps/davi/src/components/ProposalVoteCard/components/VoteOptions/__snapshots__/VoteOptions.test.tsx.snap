// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteResults matches the snapshot 1`] = `
.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 8px;
  width: 8px;
  background-color: #00F5F5;
  border-radius: 50%;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 8px;
  width: 8px;
  background-color: #FD568C;
  border-radius: 50%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-weight: 400;
  font-size: 14px;
  margin: 5px 0px 5px 0px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  margin-right: 0.5rem;
}

<div>
   
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <span
          class="c3"
          color="#00F5F5"
          size="8"
        />
      </span>
      No
    </div>
  </div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <span
          class="c4"
          color="#FD568C"
          size="8"
        />
      </span>
      Yes
    </div>
  </div>
   
</div>
`;
